### PR TITLE
Esad/refactor

### PR DESCRIPTION
**Pull request by eyusufatik**

### CODE REVIEW

File: src/service.rs [hunk 1]
```diff
@@ -32,25 +34,10 @@ pub struct BitcoinService {
     client: BitcoinNode,
     rollup_name: String,
     network: bitcoin::Network,
-    address: String,
+    address: Address<NetworkUnchecked>,
```
Comment (orkunkilic) [inline src/service.rs:37]: Why unchecked if we are forcing the network

File: src/service.rs [hunk 2]
```diff
@@ -78,7 +69,7 @@ impl BitcoinService {
     // Create a new instance of the DA service from the given configuration.
     pub fn new(config: DaServiceConfig, chain_params: RollupParams) -> Self {
         let network =
-            bitcoin::Network::from_str(&config.network.unwrap_or("regtest".to_owned())).unwrap(); // default to regtest (?)
+            bitcoin::Network::from_str(&config.network).unwrap_or(bitcoin::Network::Regtest); // default to regtest (?)
```
Comment (orkunkilic) [inline src/service.rs:73]: Let's force the network

File: src/service.rs [hunk 3]
```diff
@@ -62,13 +49,17 @@ pub struct DaServiceConfig {
...
     // taproot address that holds the funds of the sequencer
-    pub address: Option<String>,
+    // will be used as the change address for the inscribe transaction
+    pub address: String,
 
     // da private key of the sequencer
     pub sequencer_da_private_key: Option<String>,
+
+    // number of last paid fee rates to average if estimation fails
+    pub fee_rates_to_avg: usize,
```
Comment (orkunkilic) [inline src/service.rs:62]: Let's make this Option and have an enum that makes Network -> DefaultFeeRate fallback

File: src/service.rs [hunk 4]
```diff
@@ -177,27 +188,41 @@ impl DaService for BitcoinService {
...
+                let public_key = secp256k1::PublicKey::from_slice(&inscription.public_key);
+                let signature = ecdsa::Signature::from_compact(&inscription.signature);
+                let message = Message::from_hashed_data::<sha256d::Hash>(&inscription.body);
+
+                let secp = Secp256k1::new();
+                if public_key.is_ok()
+                    && signature.is_ok()
+                    && secp
+                        .verify_ecdsa(&message, &signature.unwrap(), &public_key.unwrap())
+                        .is_ok()
```
Comment (orkunkilic) [inline src/service.rs:213]: Let's move this to a helper func.

File: src/service.rs [hunk 5]
```diff
@@ -265,41 +290,67 @@ impl DaService for BitcoinService {
 
         let blob = blob.to_vec();
         let network = self.network;
-        let address = self.address.clone();
+        let address = self
+            .address
+            .clone()
+            .require_network(network)
+            .expect("Invalid network for address");
```
Comment (orkunkilic) [inline src/service.rs:297]: This should be validated in the with_client not here

File: src/service.rs [hunk 6]
```diff
@@ -265,41 +290,67 @@ impl DaService for BitcoinService {
...
+                for fee in last_fee_rates.iter() {
+                    sum += fee;
+                }
+
+                let len = last_fee_rates.len();
+                if len > 0 {
+                    sum / len as f64
+                } else {
+                    12.0
+                }
```
Comment (orkunkilic) [inline src/service.rs:341]: helper function please

File: src/service.rs [hunk 7]
```diff
@@ -265,41 +290,67 @@ impl DaService for BitcoinService {
...
             &rollup_name,
             blob,
             signature,
             public_key,
-            satpoint,
             utxos,
-            change_addresses,
-            destination_address,
+            address,
+            546,
```
Comment (orkunkilic) [inline src/service.rs:353]: Let's make this const in the file

File: src/verifier.rs [hunk 8]
```diff
@@ -73,101 +74,132 @@ impl DaVerifier for BitcoinVerifier {
...
+                        && signature.is_ok()
+                        && message.is_ok()
+                        && secp
+                            .verify_ecdsa(
+                                &message.unwrap(),
+                                &signature.unwrap(),
+                                &public_key.unwrap(),
+                            )
+                            .is_ok()
+                    {
```
Comment (orkunkilic) [inline src/verifier.rs:148]: We can use this as a helper because it is common in service also

File: src/verifier.rs [hunk 9]
```diff
@@ -176,10 +208,14 @@ impl DaVerifier for BitcoinVerifier {
...
-            .to_raw_hash()
-            .to_byte_array();
+        let root_from_inclusion = merkle_tree::calculate_root(tx_hashes.into_iter());
+
+        assert!(
+            root_from_inclusion.is_some(),
+            "merkle root couldn't be computed"
+        );
+
+        let root_from_inclusion = root_from_inclusion.unwrap().to_raw_hash().to_byte_array();
```
Comment (orkunkilic) [inline src/verifier.rs:218]: we can make this match {
    some(.)
    none
}

**Review by orkunkilic — COMMENTED** [review]
Left some comments otherwise lgtm

File: src/service.rs (hunk 1 above)
Comment (eyusufatik) [reply src/service.rs:37]: Network checked cannot be serialized so the struct has unchecked but before use it enforces the network (panics if fails)

File: src/service.rs (hunk 5 above)
Comment (eyusufatik) [reply src/service.rs:297]: see reply I left on the first comment

File: src/service.rs (hunk 1 above)
Comment (orkunkilic) [reply src/service.rs:37]: ok then store it as unchecked but make the check in with_client please not before sending

File: src/service.rs (hunk 3 above)
Comment (eyusufatik) [reply src/service.rs:62]: this is the number of last paid fee rates to use when you can't get your normal feerate

What do you mean when you say Network -> DeafualtFeeRate fallback?

Comment (orkunkilic) [reply src/service.rs:62]: DAServiceConfig is given when the rollup is started running. User has to input a fee_rates_to_avg. If this is not intended you should remove it from the config and only keep it in the struct.

If so use a default value per network to initiate this var of service struct

**Pull request merged**